### DISCUSSION TITLE
Have Android Permission Issue When Loading Assemblies

### DISCUSSION THREAD

**Discussion by CYLMos** [AvaloniaUI/Avalonia#9404] [Q&A] [answered]
I have a storage permission issue when loading assemblies.

The error in device log:
```
System.UnauthorizedAccessException: UnauthorizedAccess_IODenied_Path, /
System.IO.IOException: Permission denied
```
![Screenshot 2022-11-10 155548](https://user-images.githubusercontent.com/14320893/201033202-f8026d41-2bac-4c52-a4b5-47dec93b6b3d.png)

The exception is happened in the following code:
```
var assemblyPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location) + @"/";
var files = Directory.GetFiles(assemblyPath, Constant.PluginDllPattern); // Dead here
foreach (var file in files)
{
    try
    { 
        var assembly = Assembly.LoadFrom(file);
        // Do Somethings
    }
    catch (Exception ex)
    { // Do Somethings }
}
```

My Android device version is 10, min and max SDK version in AndroidMainfest.xml are 21 and 31.
The following is the content of my AndroidMainfest.xml
```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android" xmlns:tools="http://schemas.android.com/tools" android:installLocation="auto">
  <uses-sdk android:minSdkVersion="21" android:targetSdkVersion="31" />
  <uses-feature android:name="android.hardware.usb.host" android:required="true" />
  <uses-permission android:name="android.permission.INTERNET" />
  <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" android:maxSdkVersion="29"/>
  <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" android:maxSdkVersion="29"/>
  <uses-permission android:name="android.permission.WRITE_INTERNAL_STORAGE" android:maxSdkVersion="29"/>
  <uses-permission android:name="android.permission.READ_INTERNAL_STORAGE" android:maxSdkVersion="29"/>
  <uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE" />
  <application android:label="HelloWorld" android:icon="@drawable/Icon" android:requestLegacyExternalStorage="true" />
</manifest>
```

I'm sure I have grant the storage permission to my application but I still get crash after the error occures.
By the way, this issue is only happened in release mode. It's fine in debug mode.

My Avalonia version is 0.10.999-cibuild0023268-beta.

**Comment by emmauss** [comment, marked as answer]
Android doesn't have the concept of an executing path. Assemblies are stored in the apps private internal storage. Read more here, https://learn.microsoft.com/en-us/xamarin/android/platform/files/ .
The relevant path where assemblies are stored is `/data/user/0/package.id/files/.__override__`

**Reply by emmauss** [reply]
If the issue only occurs in Release mode, then you have AOT compilation enabled. This is the default in Release configurations for .net Android projects.
